CSP for custom domains (#3864)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/guides/custom-error-pages/main/index.md

---

> title: Customize the error pages

Reviewer: Maybe simplify to "Customize error pages"?

Reviewer 2: we need to doc they should not be using `meta` tags to customize the csp on customizable pages and we also need to test the implications of this cc @kevinroh-okta @zeusandrade-okta @bhagyaprabhakar-okta @bryanhonan-okta @rishabhchandna-okta

Reviewer 2: https://okta.slack.com/archives/CUK1VSYTZ/p1675372178689119?thread_ts=1675368555.780589&cid=CUK1VSYTZ

Author: @mihaiiacob-okta Thank you! I have another jira for it that I'm adding to this PR.

## About error page customization

When using redirect authentication, you can create a unique sign-in experience. First, set up a [custom domain](/docs/guides/custom-url-domain/) and then create a [Widget style](/docs/guides/custom-widget/main/#style-the-okta-hosted-sign-in-widget) that matches your application's look and feel.

> If an error occurs during sign-in, Okta needs to display an error page to the user. You can use the code editor to customize the error page.

Reviewer 3: ```suggestion
If an error occurs during the sign-in flow, Okta needs to display an error page to the user. You can use the code editor to customize the error page.
```

> **Note:** A custom error page only appears when an app connects to Okta by using your custom domain. Otherwise, the default Okta error page appears when a critical error occurs, or you haven't properly configured an app. See [Okta deployment models&mdash;redirect vs. embedded](/docs/concepts/redirect-vs-embedded/#redirect-authentication) for more information on redirect functionality.

### Use the Brands API

The [Brands API](/docs/reference/api/brands/) allows you to set icons, images, and colors across the following:

> - sign-in page (redirect authentication)

Reviewer 3: ```suggestion
- Sign-in page (redirect authentication)
```

> - error pages

Reviewer 3: ```suggestion
- Error pages
```

> - email templates

Reviewer 3: ```suggestion
- Email templates
```

- Okta End-User Dashboard

You don't need to set a customized Okta URL domain to use the API.

## Content Security Policy (CSP) for your custom domain

Set up a [custom domain](/docs/guides/custom-url-domain/main/) and customize your [CSP](https://content-security-policy.com/) if you also want to customize the sign-in page or error pages. CSP customizations only take effect on custom domains.

To analyze and detect potentially malicious IP addresses that seek to bypass your CSP, use [Okta ThreatInsight](https://help.okta.com/okta_help.htm?type=oie&id=ext-about-threatinsight).

### Known limitations

> * Avoid using `meta` tags to customize the CSP. `meta` tags impact the overall policy. It's easier to control CSP customizations by adding trusted origin URLs in the Admin Console. See [Multiple content security policies](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy#multiple_content_security_policies).

Reviewer 4: I would replace "trusted origin URLs" with "Trusted External Resources" because Okta has another feature called Trusted Origins.  The CSP customization UI uses the term "trusted external resources"

> * If you add too many trusted origins to your custom CSP, the HTTP header size can exceed the limit allowed by some server software. Update the default server limits or reduce the number of trusted origins.

Reviewer 4: I would replace "trusted origin URLs" with "Trusted External Resources" because Okta has another feature called Trusted Origins. The CSP customization UI uses the term "trusted external resources"

* You can have a maximum of 20 URIs.
* If you add a trusted origin URL that redirects to a different URL, you must include the redirect URL in the trusted origin list.
<!-- nonce not supported yet
### Add a nonce reference

Add a [`nonce`](https://content-security-policy.com/nonce/) reference to your HTML if you want to customize the Okta-hosted sign-in page or error pages. Without the `nonce` reference, your `script` and `style` tags don't run.

To add the `nonce` reference, include it as a variable in the code editor. See [Edit the error page](#edit-the-error-page).

Example:

```html
<style nonce="{{nonceValue}}">
```-->

### Customize the default CSP

> 1. In the Admin Console, go to **Customizations** > **Branding**.

Reviewer 4: now that multi-brand is GA, we have to include one more menu item
**Customizations** > **Branding**.>Brand (of the custom domain you are working on)

> 2. In the **Pages** panel, click **Edit** for **Error pages**.

Reviewer 4: With multi-brand this would be on "Themes" panel

3. Click **Settings**.
4. In the **Content Security Policy** panel, click **Edit**. Set the following:

   - **Trusted external resources**: Add resources to the CSP. For example: mydomain.com, *.mydomain.com, or mydomain.com/images. Click **Add**.
     > **Note:** Okta adds these resources to all fetch-directives in the CSP.

>    - **Report UI**: Enter the URI to which you want to send violation report details. The URI entered here appears in the report-uri directive of the CSP.

Reviewer 4: Report URI

>    - **Report UI**: Enter the URI to which you want to send violation report details. The URI entered here appears in the report-uri directive of the CSP.

Reviewer 4: Actually this is labelled as "Violations Report URI"

## About error page customization

When using redirect authentication, you can create a unique sign-in experience. First, set up a [custom domain](/docs/guides/custom-url-domain/) and then create a [Widget style](/docs/guides/custom-widget/main/#style-the-okta-hosted-sign-in-widget) that matches your application's look and feel.

> However, if an error happens during sign-in, Okta needs to display an error page to the user. You can also customize the error page by using the code editor.

Reviewer: I would omit 'However,'.:

"If an error occurs during sign-in, Okta displays an error page to the user. You can use the code editor to customize the error page."

- occurs is better than 'happens', imo
- we can omit needs; Okta is displaying the error page; there's no 'need' about it
- acrolinx will complain about 'by using the code editor', so swapping the order makes it more active voice

Author: Done.

> > **Note:** A custom error page only pops up when an app connects to Okta by using your custom domain. Otherwise, the default Okta error page pops up when a critical error happens or you haven't properly configured an app. See [Okta deployment models&mdash;redirect vs. embedded](/docs/concepts/redirect-vs-embedded/#redirect-authentication) for more information on redirect functionality.

Reviewer: Why "pops up" over "appears" or "is displayed"?

Author: Acrolinx, my dude. Acrolinx.

### Use the Brands API

The [Brands API](/docs/reference/api/brands/) allows you to set icons, images, and colors across the following:

- sign-in page (redirect authentication)
- error pages
- email templates
- Okta End-User Dashboard

> You don't need to set a customized Okta URL domain. See [Customize your Okta experience with the Brands API](/docs/guides/customize-themes) for more information on this feature and how to use it.

Reviewer: I might use "...for more information on using the Brands API." or similar, because the 'this feature' could be confused with "customized Okta URL domain" in the previous sentence.

Author: Removing the reference entirely as Albert wants the guide removed.

> ## Content Security Policy (CSP) for your custom domain

Reviewer: could possibly use "Content Security Policy (CSP) for custom domains"; but I don't know the layout of the devdocs, so the current version may be more appropriate

Author: sticking with this for now

Set up a [custom domain](/docs/guides/custom-url-domain/main/) and customize your [CSP](https://content-security-policy.com/) if you also want to customize the sign-in page or error pages. CSP customizations only take effect on custom domains.

To analyze and detect potentially malicious IP addresses that seek to bypass your CSP, use [Okta ThreatInsight](https://help.okta.com/okta_help.htm?type=oie&id=ext-about-threatinsight).

> **Note:** If you add too many trusted origins to your custom CSP, the HTTP header size can exceed the limit allowed by some server software. Update the default server limits or reduce the number of trusted origins.

> **Note:** You can have a maximum of 20 URIs.

> > **Note:** If you add a trusted origin URL that redirects to a different URL, you need to include the redirect URL on the trusted origin list.

Reviewer: You could replace 'need to' with 'must'

### Add a nonce reference

Add a [`nonce`](https://content-security-policy.com/nonce/) reference to your HTML if you want to customize the Okta-hosted sign-in page or error pages. Without the `nonce` reference, your `script` and `style` tags don't run.

> To add the `nonce` reference, include it as a variable in the code editor. See [Edit the error page](#edit-the-error-page).

Reviewer: would it be obvious that they'd use the code editor? If so, we could maybe drop "in the code editor".

Author: i disagree. leaving it in.

Example:

```html
<style nonce="{{nonceValue}}">
```

### Customize the default CSP

1. In the Admin Console, go to **Customizations** > **Branding**.
2. In the **Pages** panel, click **Edit** for **Error pages**.
3. Click **Settings**.
4. In the **Content Security Policy** panel, click **Edit**. Set the following:

   - **Trusted external resources**: Add resources to the CSP. For example: mydomain.com, *.mydomain.com, or mydomain.com/images. Click **Add**.
     > **Note:** Okta adds these resources to all fetch-directives in the CSP.
   - **Report UI**: Enter the URI to which you want to send violation report details. The URI entered here appears in the report-uri directive of the CSP.
   - **Enforcement**:
     - Select **Enforced** to block any resources that the CSP doesn't trust. The header is `content-security-policy`.

>      - Use **Not enforced** to leave the customized CSP in the report-only header (`content-security-policy-report-only`). Use this option for testing and validation before turning on **Enforced** mode.

Reviewer: maybe 'before enabling **Enforced**'?

Author: done. thanks.

5. Click **Save to draft**.

> 6. Make changes directly in the editor. See steps 4 and 5 in [Edit the error page](#edit-the-error-page).

Reviewer: Maybe "Make your changes in the editor."? Not sure what directly adds.

Author: done

## About error page customization

When using redirect authentication, you can create a unique sign-in experience. First, set up a [custom domain](/docs/guides/custom-url-domain/) and then create a [Widget style](/docs/guides/custom-widget/main/#style-the-okta-hosted-sign-in-widget) that matches your application's look and feel.

If an error occurs during sign-in, Okta needs to display an error page to the user. You can use the code editor to customize the error page.

> > **Note:** A custom error page only appears when an app connects to Okta by using your custom domain. Otherwise, the default Okta error page appears when a critical error happens or you haven't properly configured an app. See [Okta deployment models&mdash;redirect vs. embedded](/docs/concepts/redirect-vs-embedded/#redirect-authentication) for more information on redirect functionality.

Reviewer 5: why change occurs to happens? occurs works better IMHO
```suggestion
> **Note:** A custom error page only appears when an app connects to Okta by using your custom domain. Otherwise, the default Okta error page appears when a critical error happens, or you haven't properly configured an app. See [Okta deployment models&mdash;redirect vs. embedded](/docs/concepts/redirect-vs-embedded/#redirect-authentication) for more information on redirect functionality.
```

### Use the Brands API

The [Brands API](/docs/reference/api/brands/) allows you to set icons, images, and colors across the following:

- sign-in page (redirect authentication)
- error pages
- email templates
- Okta End-User Dashboard

You don't need to set a customized Okta URL domain to use the API.

## Content Security Policy (CSP) for your custom domain

Set up a [custom domain](/docs/guides/custom-url-domain/main/) and customize your [CSP](https://content-security-policy.com/) if you also want to customize the sign-in page or error pages. CSP customizations only take effect on custom domains.

To analyze and detect potentially malicious IP addresses that seek to bypass your CSP, use [Okta ThreatInsight](https://help.okta.com/okta_help.htm?type=oie&id=ext-about-threatinsight).

### Known limitations

> * Don't use `meta` tags to customize the CSP. `meta` tags impact the overall policy. It's easier to control CSP customizations by adding trusted origin URLs in the Admin Console.

Author: @bhagyaprabhakar-okta How's this?

Reviewer 4: I worry about the imperative tone of "Don't use meta tags".  All these days, when we did not have this feature we did say use 'meta' tags if you need to.  CC @bryanhonan-okta and @mihaiiacob-okta  for comments.

> * Don't use `meta` tags to customize the CSP. `meta` tags impact the overall policy. It's easier to control CSP customizations by adding trusted origin URLs in the Admin Console.

Author: ```suggestion
* Avoid using `meta` tags to customize the CSP. `meta` tags impact the overall policy. It's easier to control CSP customizations by adding trusted origin URLs in the Admin Console.
```

Author: @bhagyaprabhakar-okta @bryanhonan-okta @mihaiiacob-okta

Reviewer 4: LGTM

* If you add too many trusted origins to your custom CSP, the HTTP header size can exceed the limit allowed by some server software. Update the default server limits or reduce the number of trusted origins.
* You can have a maximum of 20 URIs.

> * If you add a trusted origin URL that redirects to a different URL, you must include the redirect URL on the trusted origin list.

Reviewer 5: ```suggestion
* If you add a trusted origin URL that redirects to a different URL, you must include the redirect URL in the trusted origin list.
```

The [Brands API](/docs/reference/api/brands/) allows you to set icons, images, and colors across your Okta-hosted Sign-In Widget, error pages, email templates, and Okta End-User Dashboard all at once. You don't need to set a customized Okta URL domain. See [Customize your Okta experience with the Brands API](/docs/guides/customize-themes) for more information on this feature and how to use it.

## Content Security Policy (CSP) for your custom domain

> If you have a [custom domain](/docs/guides/custom-url-domain/main/), you need to customize the default [CSP](https://content-security-policy.com/) for your org. Okta has a default CSP that can break customizations to Okta-hosted error pages.

Reviewer 6: I'm not sure this is accurate. I think only if the org has customized sign in page they might need to customize the CSP settings and only if they add custom sources. However this might be too much detail here so I'd probably be ok leaving this as is.

Author: Okay. thank you.

> **Note:** You need to have CSP for custom domains enabled in your org.

> **Note:** CSP customizations only take effect on custom domains.

### Customize the default CSP

1. In the Admin Console, go to **Customizations** > **Branding**.
2. In the **Pages** panel, click **Edit** for **Error pages**.
3. Click **Settings**.
4. In the **Content Security Policy** panel, click **Edit**. Set the following:
   - **Trusted external resources**: Add resources to the CSP. For example: mydomain.com, *.mydomain.com, or https://mydomain.com/images. Click **Add**.
     > **Note:** Okta adds these resources to all fetch-directives in the CSP.
   - **Report UI**: Enter the URI to which you want to send violation report details. The URI entered here appears in the report-uri directive of the CSP.
   - **Enforcement**:
     - Select **Enforced** to block any resources that the CSP doesn't trust.
     - Use **Not enforced** to leave the customized CSP in the report-only header. Use this option for testing and validation before turning on **Enforced** mode.

> 5. Click **Save to draft**.

Reviewer 6: please add steps for previewing their changes and publishing them. like step 4 and 5 in the section below

Author: done.

> **Note:** If you add too many trusted origins to your custom CSP, the HTTP header size can exceed the limit allowed by some server software. You may need to update the default server limits or reduce the number of trusted origins in Okta.

> **Note:** If you add a trusted origin URL that redirects to a different URL, you need to include the redirect URL on the trusted origin list.

### Customize the default CSP

1. In the Admin Console, go to **Customizations** > **Branding**.
2. In the **Pages** panel, click **Edit** for **Error pages**.
3. Click **Settings**.
4. In the **Content Security Policy** panel, click **Edit**. Set the following:
   - **Trusted external resources**: Add resources to the CSP. For example: mydomain.com, *.mydomain.com, or https://mydomain.com/images. Click **Add**.
     > **Note:** Okta adds these resources to all fetch-directives in the CSP.
   - **Report UI**: Enter the URI to which you want to send violation report details. The URI entered here appears in the report-uri directive of the CSP.
   - **Enforcement**:
     - Select **Enforced** to block any resources that the CSP doesn't trust.

>      - Use **Not enforced** to leave the customized CSP in the report-only header. Use this option for testing and validation before turning on **Enforced** mode.

Reviewer 2: maybe refer to the header as `content-security-policy-report-only` when meaning report-only header and `content-security-policy` when meaning enforce header

Author: done

## Content Security Policy (CSP) for your custom domain

If you have a [custom domain](/docs/guides/custom-url-domain/main/), you need to customize the default [CSP](https://content-security-policy.com/) for your org. Okta has a default CSP that can break customizations to Okta-hosted error pages. CSP customizations only take effect on custom domains.

To analyze and detect potentially malicious IP addresses that seek to bypass your CSP, use [Okta ThreatInsight](https://help.okta.com/okta_help.htm?type=oie&id=ext-about-threatinsight).

> **Note:** If you add too many trusted origins to your custom CSP, the HTTP header size can exceed the limit allowed by some server software. You may need to update the default server limits or reduce the number of trusted origins in Okta.

> **Note:** If you add a trusted origin URL that redirects to a different URL, you need to include the redirect URL on the trusted origin list.

> ### Include a nonce reference

Reviewer 7: empty section

### Customize the default CSP

1. In the Admin Console, go to **Customizations** > **Branding**.
2. In the **Pages** panel, click **Edit** for **Error pages**.
3. Click **Settings**.

> 4. In the **Content Security Policy** panel, click **Edit**. Set the following:

Author: Done!

| <span v-pre>`{{bgImageUrl}}`</span> | The URL to the background image configured for your application. You can change this image by using the [Sign-in Configuration](https://help.okta.com/okta_help.htm?type=oie&id=ext-branding-set-theme) option, but this changes the background image in all instances where the variable is used, including your custom sign-in page. If you want to change only the background image for your custom error pages, include the URL to the image instead of the variable. |
| <span v-pre>`{{orgLogo}}`</span> | The logo image that has been configured for your application. You can change this logo by using the [Sign-in Configuration](https://help.okta.com/okta_help.htm?type=oie&id=ext-branding-set-theme) option, but this changes the org logo in all instances where the variable is used, including your custom sign-in page. If you want to change only the logo image for your custom error pages, include the URL to the image instead of the variable. |
| <span v-pre>`{{{errorDescription}}}`</span> | A detailed description of the error |
| <span v-pre>`{{back}}`</span> | The text "Go to Homepage". When the user clicks the button, they’re returned to the sign-in page. |

> | <span v-pre>`{{technicalDetails}}`</span> | Any additional messaging, if the error code has any. Here are sample technical details for an error code:</br>"If you’re using custom expressions like `\{0}`, make sure that the field `customField` is present in the user profile. Please review your attribute list and make the appropriate change."</br>See [Okta Error Codes](/docs/reference/error-codes/#okta-error-codes-listed-by-error-code). |

Reviewer 5: ```suggestion
| <span v-pre>`{{technicalDetails}}`</span> | Any additional messaging, if the error code has any. Here are sample technical details for an error code:</br>"If you’re using custom expressions like `\{0}`, make sure that the field `customField` is present in the user profile. Review your attribute list and make the appropriate change."</br>See [Okta Error Codes](/docs/reference/error-codes/#okta-error-codes-listed-by-error-code). |
```

</head>
```

> ## About the Content Security Policy (CSP) for custom domains

Author: ```suggestion
## About the Content Security Policy (CSP) for your custom domain
```

> If you have a [custom domain](/docs/guides/custom-url-domain/main/), you need to customize [CSP](https://content-security-policy.com/) for your org. Okta has a default CSP that can break customizations to Okta-hosted error pages.

Author: ```suggestion
If you have a [custom domain](/docs/guides/custom-url-domain/main/), you need to customize the default [CSP](https://content-security-policy.com/) for your org. Okta has a default CSP that can break customizations to Okta-hosted error pages.
```

To analyze and detect potentially malicious IP addresses that seek to bypass your CSP, use [Okta ThreatInsight](https://help.okta.com/okta_help.htm?type=oie&id=ext-about-threatinsight).

> **Note:** You need to have CSP for custom domains enabled in your org.

> **Note:** CSP customizations only take effect on custom domains.

> ### Configure CSP for custom domains

Author: ```suggestion
### Customize the default CSP
```

1. In the Admin Console, go to **Customizations** > **Branding**.
2. In the **Pages** panel, click **Edit** for **Error pages**.
3. Click **Settings**.
4. In the **Content Security Policy** panel, click **Edit**. Set the following:

>    - **Trusted external resources**: Add resources to the CSP. For example: mydomain.com, *.mydomain.com, or https://mydomain.com/images. Click **Add**.

Reviewer 4: these resources are added to all fetch-directives that Okta configures in the CSP

Author: Added.

>    - **Report UI**: Enter the URI to which you want to send report details.

Reviewer 4: enter the URI to which you want to send violation report details.  The URI entered here will appear in the report-uri directive of the CSP

Author: Added.

>    - **Enforcement**: Select **Enforced** to block any resources that the CSP doesn't trust. Use **Not enforced** for testing only.

Reviewer 4: Not Enforced leaves the config in CSP Report-only header.  It can be used for testing and validation before turning on in enforced mode.

Author: Updated.
